### PR TITLE
parameters_groupspecific [WIP]

### PR DESCRIPTION
**Pull request by DominiqueMakowski**
#482

As underlined by @bwiernik through #484 we can either by default return the "raw" random parameters or the sum with their fixed one. I'm not sure what the best default behaviour is, but we definitely should have an argument to get one or the other. 

- What would the name of that argument be?
- What is the technical name of that sum as opposed to the raw random params? 
- Is there a way to get that "sum" by taking into account the uncertainty of the fixed + of the random?

### CODE REVIEW

**Comment by DominiqueMakowski** [comment]
I've no idea how to deal with nested/crossed RE

**Comment by mattansb** [comment]
>I've no idea how to deal with nested/crossed RE


Where are you stuck?

**Comment by DominiqueMakowski** [comment]
I'll try a bit more things first!

**Comment by DominiqueMakowski** [comment]
Here's what I have for now; I'm not sure what to think of it:

Essentially the main, default output of a groupspecific function is to return data of the same shape as the input data, so it can eventually be directly merged with it. However, there is also the `summary()` method that can be used to get the "long" format that has one row per parameter ("long" and "wide" are tricky here 'cause in the end the wide format ends up being longer because of rows duplication).  In the "wide" format (the default one), all the parameters for each group are put on the same line and are duplicated at each level (e.g., `Species | Species_Coefficient_Intercept | Species_Coefficient_Sepal.Length`, in which the parameters corresponding to the intercept and the slope of the Species are on the same line - hence "wide" format).


Now there's the problem of nested effects, I don't know how to address it:

https://github.com/easystats/parameters/blob/8d6cbc2752b24173064bb36a0d22db13b82ba797/R/parameters_groupspecific.R#L49-L50

Because here I extract the random terms but the nested ones do not exist as separate columns in the original dataframe so it errors

What do you think?

``` r
library(lme4)
#> Loading required package: Matrix
library(parameters)

data <- lme4::sleepstudy

# Random intercept
model <- lmer(Reaction ~ Days + (1 | Subject), data = data)
random <- parameters_groupspecific(model)
head(random, 12)
#>    Subject Subject_Coefficient_Intercept Subject_SE_Intercept
#> 1      308                      40.78371             9.475668
#> 2      308                      40.78371             9.475668
#> 3      308                      40.78371             9.475668
#> 4      308                      40.78371             9.475668
#> 5      308                      40.78371             9.475668
#> 6      308                      40.78371             9.475668
#> 7      308                      40.78371             9.475668
#> 8      308                      40.78371             9.475668
#> 9      308                      40.78371             9.475668
#> 10     308                      40.78371             9.475668
#> 11     309                     -77.84955             9.475668
#> 12     309                     -77.84955             9.475668
#>    Subject_CI_low_Intercept Subject_CI_high_Intercept
#> 1                  22.21174                  59.35568
#> 2                  22.21174                  59.35568
#> 3                  22.21174                  59.35568
#> 4                  22.21174                  59.35568
#> 5                  22.21174                  59.35568
#> 6                  22.21174                  59.35568
#> 7                  22.21174                  59.35568
#> 8                  22.21174                  59.35568
#> 9                  22.21174                  59.35568
#> 10                 22.21174                  59.35568
#> 11                -96.42152                 -59.27759
#> 12                -96.42152                 -59.27759
summary(random)
#>      Group Level   Parameter Coefficient       SE   CI      CI_low    CI_high
#> 1  Subject   308 (Intercept)   40.783710 9.475668 0.95  22.2117416  59.355678
#> 2  Subject   309 (Intercept)  -77.849554 9.475668 0.95 -96.4215221 -59.277586
#> 3  Subject   310 (Intercept)  -63.108567 9.475668 0.95 -81.6805357 -44.536599
#> 4  Subject   330 (Intercept)    4.406442 9.475668 0.95 -14.1655261  22.978410
#> 5  Subject   331 (Intercept)   10.216189 9.475668 0.95  -8.3557791  28.788157
#> 6  Subject   332 (Intercept)    8.221238 9.475668 0.95 -10.3507303  26.793206
#> 7  Subject   333 (Intercept)   16.500494 9.475668 0.95  -2.0714741  35.072462
#> 8  Subject   334 (Intercept)   -2.996981 9.475668 0.95 -21.5689494  15.574987
#> 9  Subject   335 (Intercept)  -45.282127 9.475668 0.95 -63.8540954 -26.710159
#> 10 Subject   337 (Intercept)   72.182686 9.475668 0.95  53.6107179  90.754654
#> 11 Subject   349 (Intercept)  -21.196249 9.475668 0.95 -39.7682172  -2.624281
#> 12 Subject   350 (Intercept)   14.111363 9.475668 0.95  -4.4606049  32.683332
#> 13 Subject   351 (Intercept)   -7.862221 9.475668 0.95 -26.4341894  10.709747
#> 14 Subject   352 (Intercept)   36.378425 9.475668 0.95  17.8064571  54.950394
#> 15 Subject   369 (Intercept)    7.036381 9.475668 0.95 -11.5355873  25.608349
#> 16 Subject   370 (Intercept)   -6.362703 9.475668 0.95 -24.9346713  12.209265
#> 17 Subject   371 (Intercept)   -3.294273 9.475668 0.95 -21.8662411  15.277695
#> 18 Subject   372 (Intercept)   18.115747 9.475668 0.95  -0.4562215  36.687715




# Random slope and intercept
model <- lmer(Reaction ~ Days + (1 + Days  | Subject), data = data)
random <- parameters_groupspecific(model)
head(random, 12)
#>    Subject Subject_Coefficient_Intercept Subject_SE_Intercept
#> 1      308                      2.258551             12.07086
#> 2      308                      2.258551             12.07086
#> 3      308                      2.258551             12.07086
#> 4      308                      2.258551             12.07086
#> 5      308                      2.258551             12.07086
#> 6      308                      2.258551             12.07086
#> 7      308                      2.258551             12.07086
#> 8      308                      2.258551             12.07086
#> 9      308                      2.258551             12.07086
#> 10     308                      2.258551             12.07086
#> 11     309                    -40.398738             12.07086
#> 12     309                    -40.398738             12.07086
#>    Subject_CI_low_Intercept Subject_CI_high_Intercept Subject_Coefficient_Days
#> 1                 -21.39989                  25.91700                 9.198976
#> 2                 -21.39989                  25.91700                 9.198976
#> 3                 -21.39989                  25.91700                 9.198976
#> 4                 -21.39989                  25.91700                 9.198976
#> 5                 -21.39989                  25.91700                 9.198976
#> 6                 -21.39989                  25.91700                 9.198976
#> 7                 -21.39989                  25.91700                 9.198976
#> 8                 -21.39989                  25.91700                 9.198976
#> 9                 -21.39989                  25.91700                 9.198976
#> 10                -21.39989                  25.91700                 9.198976
#> 11                -64.05718                 -16.74029                -8.619681
#> 12                -64.05718                 -16.74029                -8.619681
#>    Subject_SE_Days Subject_CI_low_Days Subject_CI_high_Days
#> 1         2.304839            4.681574            13.716377
#> 2         2.304839            4.681574            13.716377
#> 3         2.304839            4.681574            13.716377
#> 4         2.304839            4.681574            13.716377
#> 5         2.304839            4.681574            13.716377
#> 6         2.304839            4.681574            13.716377
#> 7         2.304839            4.681574            13.716377
#> 8         2.304839            4.681574            13.716377
#> 9         2.304839            4.681574            13.716377
#> 10        2.304839            4.681574            13.716377
#> 11        2.304839          -13.137082            -4.102279
#> 12        2.304839          -13.137082            -4.102279

# Multiple random groups
model <- lmer(mpg ~ wt + (1 | gear) + (1 | carb), data = mtcars)
random <- parameters_groupspecific(model)
head(random, 12)
#>    carb gear gear_Coefficient_Intercept gear_SE_Intercept gear_CI_low_Intercept
#> 1     1    3                 -0.5433673         0.7080738            -1.9311665
#> 2     1    4                  1.2241982         0.7657549            -0.2766538
#> 3     1    4                  1.2241982         0.7657549            -0.2766538
#> 4     1    4                  1.2241982         0.7657549            -0.2766538
#> 5     1    3                 -0.5433673         0.7080738            -1.9311665
#> 6     1    3                 -0.5433673         0.7080738            -1.9311665
#> 7     1    4                  1.2241982         0.7657549            -0.2766538
#> 8     2    3                 -0.5433673         0.7080738            -1.9311665
#> 9     2    3                 -0.5433673         0.7080738            -1.9311665
#> 10    2    3                 -0.5433673         0.7080738            -1.9311665
#> 11    2    4                  1.2241982         0.7657549            -0.2766538
#> 12    2    3                 -0.5433673         0.7080738            -1.9311665
#>    gear_CI_high_Intercept carb_Coefficient_Intercept carb_SE_Intercept
#> 1               0.8444319                  0.5499008         0.6750288
#> 2               2.7250501                  0.5499008         0.6750288
#> 3               2.7250501                  0.5499008         0.6750288
#> 4               2.7250501                  0.5499008         0.6750288
#> 5               0.8444319                  0.5499008         0.6750288
#> 6               0.8444319                  0.5499008         0.6750288
#> 7               2.7250501                  0.5499008         0.6750288
#> 8               0.8444319                  0.2925632         0.6385148
#> 9               0.8444319                  0.2925632         0.6385148
#> 10              0.8444319                  0.2925632         0.6385148
#> 11              2.7250501                  0.2925632         0.6385148
#> 12              0.8444319                  0.2925632         0.6385148
#>    carb_CI_low_Intercept carb_CI_high_Intercept
#> 1             -0.7731312               1.872933
#> 2             -0.7731312               1.872933
#> 3             -0.7731312               1.872933
#> 4             -0.7731312               1.872933
#> 5             -0.7731312               1.872933
#> 6             -0.7731312               1.872933
#> 7             -0.7731312               1.872933
#> 8             -0.9589028               1.544029
#> 9             -0.9589028               1.544029
#> 10            -0.9589028               1.544029
#> 11            -0.9589028               1.544029
#> 12            -0.9589028               1.544029
```

<sup>Created on 2021-05-09 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by bwiernik** [comment]
Let's take a step back. There are two different use cases here. 

1. Extract the group-specific parameter estimates. That is, the estimated slopes/intercepts for each level of a grouping factor. This is what is returned by the `coef()` function. brms and I think nlme will include SEs and intervals with these estimates. glmmTMB May in the future. lme4 won't. These should be returned in the same format as the "random" when `group_level` is TRUE. 

2. Case-level estimates of random effects augmented to the original data frame. This should be something like `get_*` I think because it's like `get_predicted`.

**Comment by DominiqueMakowski** [comment]
I'm not sure I follow you (I have troubles with the "augmented to" expression); is it correct that:

- In a mixed model, when one is interested in the individual groups' "effects" or variability, one can run `ranef()` and get the individual coefficients relative of the individuals relative to the "main" (fixed) effect. This is also returned by `parameters::parameters(model, effects = "random", group_level = TRUE)`. 
- These random effects could also be reshaped to match the input shape of the random factors, i.e., essentially duplicated for each row of the same random level. This is mostly for convenience to facilitate re-usability of estimates and don't provide any new insights.
- Instead of the deviation from the fixed effect, one can also get directly the intercept/slope with `coef()`, which corresponds to the random coef + the fixed one.
-This information is presented in the same shape as `ranef()`, and thus could have the same shape as `parameters::parameters(model, effects = "random", group_level = TRUE)`. The only difference would be that for `lme4` we would miss some information (no SE etc.).
- These "full" random effects could be also reshaped to the same format as the input random factors.
- Both the relative and the full could be of interest when it comes to study interindividual variability

In other words, we have 2x2 categories: the first one [relative vs. full (?)] is whether the values correspond to the coefficients relative to the fixed (ranef) or the full (the sum; coef), the second one is the shape of the output, either 1 row per parameter (as returned by `parameters::parameters(model, effects = "random", group_level = TRUE)`) or reshaped to match the input random dataframe.

**Comment by mattansb** [comment]
Why are we returning the data shaped to the original data? Seems like a massive waste of rows. If someone wants to append these data back to their original df (why?) they can use a simple `merge()`, there really is no need for that to happen on our end.

**Comment by DominiqueMakowski** [comment]
> Why are we returning the data shaped to the original data?

I suppose that's true too 😅

**Comment by bwiernik** [comment]
See the problem with that here https://github.com/easystats/parameters/issues/482#issuecomment-833604581

**Comment by DominiqueMakowski** [comment]
I'm almsot done in adapting @bwiernik's code to deal with nested random factors using base R, it now works BUT

``` r
library(lme4)
#> Loading required package: Matrix
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(parameters)


# Brenton's example
tidyverse_solution <- function(model, data) {
  reshaper <- function(x) {
    wider <-
      tidyr::pivot_wider(
        x,
        id_cols = c("Parameter", "Coefficient", "SE"),
        names_from = "Group",
        values_from = "Level"
      )
    if (any(stringr::str_detect(colnames(wider), ":"))) {
      column <- colnames(wider)[stringr::str_detect(colnames(wider), ":")]
      wider <- tidyr::separate(
        wider,
        col = column,
        into = strsplit(column, ":")[[1]],
        sep = ":",
        remove = TRUE
      )
    }
    wider
  }


  re_est <- parameters(model, effects = "random", group_level = TRUE) %>%
    as.data.frame() %>%
    select(Parameter, Group, Level, Coefficient, SE) %>%
    mutate(Parameter = paste(Parameter, Group, sep= "_")) %>%
    split(., .$Group) %>%
    lapply(reshaper)

  augdata <- lapply(re_est, function(x) full_join(data, x, by = colnames(x)[-c(1:3)])) %>%
    Reduce(rbind, .) %>%
    tidyr::pivot_wider(id_cols = -c(Parameter, Coefficient, SE),
                       names_from = "Parameter",
                       values_from = c("Coefficient", "SE")) %>%
    as.data.frame()
  augdata
}



data <- lme4::sleepstudy

# Simple 1
model <- lmer(Reaction ~ Days + (1 | Subject), data = data)

random <- parameters_groupspecific(model)
random2 <- tidyverse_solution(model)
#> Error in full_join(data, x, by = colnames(x)[-c(1:3)]): argument "data" is missing, with no default
all(random == random2[c(3:5)])
#> Error in eval(expr, envir, enclos): object 'random2' not found

# Simple 2
model <- lmer(Reaction ~ Days + (1 + Days  | Subject), data = data)

random <- parameters_groupspecific(model)
random2 <- tidyverse_solution(model)
#> Error in full_join(data, x, by = colnames(x)[-c(1:3)]): argument "data" is missing, with no default
all(random == random2[c(3, 4, 6, 5, 7)])
#> Error in eval(expr, envir, enclos): object 'random2' not found


# Complex 1
model <- lmer(mpg ~ wt + (1 | gear) + (1 | carb), data = mtcars)

random <- parameters_groupspecific(model)
random2 <- tidyverse_solution(model, data = mtcars)  # Fails
#> Error: Can't join on `x$carb` x `y$carb` because of incompatible types.
#> i `x$carb` is of type <double>>.
#> i `y$carb` is of type <character>>.


# Complex 2
set.seed(33)
data$grp <- sample(letters[1:5], size = 180, replace = TRUE)
data$subgrp <- NA
for (i in letters[1:5]) {
  filter_group <- data$grp == i
  data$subgrp[filter_group] <- sample(LETTERS, size = sum(filter_group), replace = TRUE)
}
model <- lmer(Reaction ~ Days + (1 | grp / subgrp) + (1 | Subject), data = data)

random <- parameters_groupspecific(model)
random2 <- tidyverse_solution(model, data = data)
# all(random[c(1, 2, 3, 4, 5, 8, 9)] == random2[c(4, 3, 5, 7, 10, 6, 11)])

head(data[c("grp", "Subject", "subgrp")])
#>   grp Subject subgrp
#> 1   b     308      E
#> 2   d     308      E
#> 3   a     308      U
#> 4   b     308      E
#> 5   d     308      K
#> 6   a     308      I
head(random[c("grp", "Subject", "subgrp")])
#>   grp Subject subgrp
#> 1   b     308      E
#> 2   b     308      E
#> 3   b     332      O
#> 4   b     350      X
#> 5   b     308      X
#> 6   b     350      A
# head(random2[c("grp", "Subject", "subgrp")])
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>


It messes up the order upon merging, despite having set:

https://github.com/easystats/parameters/blob/087a1819db6c80b1ea444930c42699c4939de0ce/R/parameters_groupspecific.R#L71

(it's at this line where it changes the order of the output - I want to keep the order of the first argument)

@strengejacke can we merge two DFs without sorting / changing the left part?

**Comment by strengejacke** [comment]
I'll look into this... I sometimes work with an ID variable that I use for sorting rows after merging

**Comment by DominiqueMakowski** [comment]
@strengejacke if you find time could you take a look at the `merge`-related (dis)order issue - that it messes up the order of rows? Worst case scenario, would it solve the problem to have a safer `data_merge()` function?

**Comment by DominiqueMakowski** [comment]
Just to clarify, the current merge function somehow transforms the order, which is not the one of the original data anymore:

``` r
set.seed(33)

data <- lme4::sleepstudy
data$grp <- sample(letters[1:5], size = 180, replace = TRUE)
data$subgrp <- NA
for (i in letters[1:5]) {
  filter_group <- data$grp == i
  data$subgrp[filter_group] <- sample(LETTERS, size = sum(filter_group), replace = TRUE)
}

model <- lme4::lmer(Reaction ~ Days + (1 | grp / subgrp) + (1 | Subject), data = data)

head(parameters::parameters_groupspecific(model)[c("grp", "Subject", "subgrp")])
#>   grp Subject subgrp
#> 1   b     308      E
#> 2   b     308      E
#> 3   b     332      O
#> 4   b     350      X
#> 5   b     308      X
#> 6   b     350      A

head(data[c("grp", "Subject", "subgrp")])
#>   grp Subject subgrp
#> 1   b     308      E
#> 2   d     308      E
#> 3   a     308      U
#> 4   b     308      E
#> 5   d     308      K
#> 6   a     308      I
```

<sup>Created on 2021-05-26 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by bwiernik** [comment]
Yes, merge sorts the output by the key. It's one if it's annoying "features" that the dplyr join functions don't share. 

@DominiqueMakowski you mentioned somewhere that perhaps this should go in modelbased instead?

**Comment by DominiqueMakowski** [comment]
> you mentioned somewhere that perhaps this should go in modelbased instead?

Yeah, but I'm quite torn. `estimate_individual()` would be maybe neater, but at the same it's also parameters? But maybe it would gain more visibility over there? I don't know, what do you think?

File: R/parameters_groupspecific.R [hunk 1]
```diff
@@ -74,15 +77,19 @@ parameters_groupspecific <- function(model, indices = c("Coefficient", "SE"), ..
...
+    if (grepl(":", random_group)) {
       groups <- as.data.frame(t(sapply(strsplit(wide[[random_group]], ":"), function(x) as.data.frame(t(x)))))
       names(groups) <- unlist(strsplit(random_group, ":"))
       wide <- cbind(groups, wide)
       wide[random_group] <- NULL
       random_group <- names(groups)
     }
 
+    # merge and sort
+    df_random[["__sort_id"]] <- 1:nrow(df_random)
```
Comment (DominiqueMakowski) [inline R/parameters_groupspecific.R:89]: we really should add a `data_merge()` function in datawiz (@IndrajeetPatil) to streamline this kind of stuff

**Comment by bwiernik** [comment]
I would say the output that gives the group-specific coefficients and their intervals should go in parameters. The output that merges these with the original data should go in modelbased

**Comment by DominiqueMakowski** [comment]
ok, makes sense. will move it to modelbased then, it will be more salient there
